File: 프로그래머스/KaKao/양궁대회.py
# 2022 KAKAO BLIND RECRUITMENT 양궁대회
from itertools import combinations_with_replacement


def getScore(lion, aporch):
    aScore, lScore = 0, 0

    for i in range(11):
        if lion[i] == 0 and aporch[i] == 0:
            continue

        if lion[i] > aporch[i]:
            lScore += (10 - i)
        else:
            aScore += (10 - i)

    return (lScore, aScore)


def solution(n, info):
    answer = []
    maxDiff = -1  # 가장 낮은 점수를 더 많이 맞힌 경우를 위해

    for combi in list(combinations_with_replacement(range(11), n)):
        lionInfo = [0] * 11

        for c in combi:
            lionInfo[10-c] += 1

        lScore, aScore = getScore(lionInfo, info)

        if lScore > aScore and lScore - aScore > maxDiff:
            maxDiff = lScore - aScore
            answer = lionInfo

    return answer if answer else [-1]

File: 프로그래머스/KaKao/양궁대회.swift
// 2022 KAKAO BLIND RECRUITMENT 양궁대회
// https://school.programmers.co.kr/learn/courses/30/lessons/92342
import Foundation

func solution(_ n:Int, _ info:[Int]) -> [Int] {
    var answer: [Int] = [-1]
    var maxDiff: Int = -1
    let totalCase = combination(Array(0...10), n)
    
    for arrows in totalCase {
        let lion = getLionInfo(arrows)
        let scores: (apeach: Int, lion: Int) = calculateScore(info, lion)
        
        if scores.apeach < scores.lion {
            if maxDiff < scores.lion - scores.apeach {
                maxDiff = scores.lion - scores.apeach
                answer = lion
            }
        }
    }
    return answer
}

func combination(_ n: [Int], _ r: Int) -> [[Int]] {
    var ret: [[Int]] = []
    
    func DFS(_ com: [Int], _ cur: Int, _ depth: Int) {
        if depth == r {
            ret.append(com)
            return
        }
        
        for i in cur..<n.count {            
            DFS(com + [n[i]], i, depth+1)
        }
    }
    
    DFS([], 0, 0)
    return ret
}

func getLionInfo(_ arrows: [Int]) -> [Int] {
    arrows.reduce(into: Array(repeating: 0, count: 11)) { $0[10-$1] += 1 }
}

func calculateScore(_ apeach: [Int], _ lion: [Int]) -> (Int, Int) {
    var ret: (Int, Int) = (0, 0)
    
    for i in 0...10 {
        if apeach[i] == 0 && lion[i] == 0 { continue }
        
        if apeach[i] < lion[i] {
            ret.1 += (10-i)
        } else {
            ret.0 += (10-i)
        }
    }
    
    return ret
}